File: windows/Marquee.Windows/Services/Updater.cs
using System.ComponentModel;
using System.Diagnostics;
using Microsoft.Win32;
using CommunityToolkit.Mvvm.ComponentModel;
using Marquee.Core;
using Marquee.Core.Updates;
using Microsoft.UI.Dispatching;

namespace Marquee.Windows.Services;

/// <summary>Where the updater is, from a check to the installer taking over.</summary>
public enum UpdatePhase
{
    Idle,
    Checking,
    UpToDate,

    /// <summary><see cref="Updater.Update"/> is newer than this app.</summary>
    Available,

    /// <summary><see cref="Updater.Progress"/> of the installer.</summary>
    Downloading,

    /// <summary>The installer is running; Marquee is about to quit, and reopens on the new version.</summary>
    Installing,

    /// <summary><see cref="Updater.ErrorMessage"/> says why.</summary>
    Failed,
}

/// <summary>
/// Keeps this PC on the newest Marquee, the Mac app's <c>Updater</c> for
/// Windows: checks GitHub's latest release ten seconds after launch and then
/// daily (and whenever Settings › About's "Check for updates" is clicked),
/// and on "Update" downloads <c>Marquee-Setup.exe</c>, checks its size and
/// SHA-256 against the release, runs it silently and quits. The installer
/// (windows/installer/Marquee.iss) replaces the app in place and, given
/// <c>/relaunch=1</c>, opens the new version. Lives on the UI thread; every
/// property changes there.
/// </summary>
public sealed partial class Updater : ObservableObject
{
    public static readonly TimeSpan FirstCheckDelay = TimeSpan.FromSeconds(10);
    public static readonly TimeSpan CheckInterval = TimeSpan.FromHours(24);

    /// <summary>
    /// Inno Setup's switches: no wizard and no message boxes, close anything
    /// still holding the app's files but don't let Inno reopen it (the
    /// script's own <c>/relaunch=1</c> step does, exactly once).
    /// </summary>
    public const string InstallerArguments =
        "/VERYSILENT /SUPPRESSMSGBOXES /NORESTART /CLOSEAPPLICATIONS /NORESTARTAPPLICATIONS /relaunch=1";

    private readonly UpdateService service;
    private readonly DispatcherQueue dispatcher;
    private DispatcherQueueTimer? timer;
    private CancellationTokenSource? installCancellation;

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(IsBusy), nameof(IsInstalling), nameof(StatusText), nameof(CanCheck), nameof(CanInstall), nameof(IsProgressIndeterminate), nameof(HasFailed), nameof(ShowsManualDownload))]
    private UpdatePhase phase = UpdatePhase.Idle;

    /// <summary>The newest release, while it's newer than this app.</summary>
    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(ShowsUpdate), nameof(StatusText), nameof(UpdateLabel), nameof(CanCheck), nameof(CanInstall), nameof(ReleasePage), nameof(ShowsManualDownload))]
    private AvailableUpdate? update;

    /// <summary>
    /// The newest published release, whatever this app's version: what the
    /// server is compared with in Settings › About. Null until a check answers.
    /// </summary>
    [ObservableProperty]
    private AppVersion? latestRelease;

    /// <summary>0 to 1 while downloading.</summary>
    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(StatusText), nameof(ProgressPercent))]
    private double progress;

    /// <summary>Why the last check or install failed; null otherwise.</summary>
    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(StatusText))]
    private string? errorMessage;

    public Updater(DispatcherQueue dispatcher, UpdateService? service = null)
    {
        this.dispatcher = dispatcher;
        this.service = service ?? new UpdateService();
    }

    /// <summary>This build's version, "0.30.0".</summary>
    public string CurrentVersion => AppInfo.Version;

    public bool IsBusy => Phase is UpdatePhase.Checking or UpdatePhase.Downloading or UpdatePhase.Installing;

    public bool IsInstalling => Phase is UpdatePhase.Downloading or UpdatePhase.Installing;

    /// <summary>A newer Marquee is known (the rail's update button shows).</summary>
    public bool ShowsUpdate => Update != null;

    /// <summary>"Check for updates": nothing newer is known and nothing is under way.</summary>
    public bool CanCheck => Update == null && !IsBusy;

    /// <summary>"Update": a newer release is known and isn't already downloading.</summary>
    public bool CanInstall => Update != null && !IsBusy && IsInstalledCopy;

    /// <summary>
    /// Whether this is the copy the installer installed (its uninstall entry
    /// points at this folder). Otherwise the installer would update that copy,
    /// not this one, which would then offer the same update every day; a dev
    /// build or a copied folder gets "Download manually" instead.
    /// </summary>
    public bool IsInstalledCopy { get; } = DetectInstalledCopy();

    /// <summary>"Download manually": after a failure, or when this copy can't update itself.</summary>
    public bool ShowsManualDownload => HasFailed || (Update != null && !IsInstalledCopy);

    public bool HasFailed => Phase == UpdatePhase.Failed;

    /// <summary>The bar has no percentage once the installer is running.</summary>
    public bool IsProgressIndeterminate => Phase == UpdatePhase.Installing;

    /// <summary>0 to 100, for the progress bar.</summary>
    public double ProgressPercent => Progress * 100;

    /// <summary>"Update to Marquee 0.31.0".</summary>
    public string UpdateLabel => Update is { } available ? $"Update to Marquee {available.Version}" : "Update Marquee";

    /// <summary>The one line Settings › About shows under the version.</summary>
    public string StatusText => Phase switch
    {
        UpdatePhase.Checking => "Checking for updates…",
        UpdatePhase.UpToDate => "You're up to date.",
        UpdatePhase.Available when Update is { } available && !IsInstalledCopy =>
            $"Marquee {available.Version} is available. This copy of Marquee isn't the installed one, so download the new version from GitHub.",
        UpdatePhase.Available when Update is { } available => $"Marquee {available.Version} is available. You have {CurrentVersion}.",
        UpdatePhase.Downloading => $"Downloading Marquee {Update?.Version}… {(int)Math.Round(Progress * 100)}%",
        UpdatePhase.Installing => "Installing. Marquee will close and reopen by itself.",
        UpdatePhase.Failed => ErrorMessage ?? "Marquee couldn't update.",
        _ => "Marquee checks for updates once a day.",
    };

    /// <summary>The release to read about or download by hand.</summary>
    public Uri ReleasePage => Update?.ReleasePage ?? UpdateService.ReleasesPage;

    /// <summary>The first check in <see cref="FirstCheckDelay"/>, then one a day.</summary>
    public void Start()
    {
        if (timer != null)
        {
            return;
        }
        timer = dispatcher.CreateTimer();
        timer.IsRepeating = false;
        timer.Interval = FirstCheckDelay;
        timer.Tick += OnTimerTick;
        timer.Start();
        // Nothing is downloading yet, so installers left by earlier updates go.
        _ = Task.Run(DeleteOldInstallers);
    }

    private async void OnTimerTick(DispatcherQueueTimer sender, object args)
    {
        sender.Stop();
        await CheckAsync();
        // A release caught in the few minutes before CI attaches its
        // downloads: look again soon rather than tomorrow.
        sender.Interval = Phase == UpdatePhase.Failed && lastCheckKind == UpdateErrorKind.NoDownload
            ? NoDownloadRetry
            : CheckInterval;
        sender.Start();
    }

    /// <summary>After a check found a newer release with no Windows download yet.</summary>
    public static readonly TimeSpan NoDownloadRetry = TimeSpan.FromMinutes(15);

    private UpdateErrorKind? lastCheckKind;

    private static string DownloadFolder => Path.Combine(Path.GetTempPath(), "Marquee-update");

    private static void DeleteOldInstallers()
    {
        try
        {
            if (Directory.Exists(DownloadFolder))
            {
                Directory.Delete(DownloadFolder, recursive: true);
            }
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException)
        {
            // In use or locked; the next launch tries again.
        }
    }

    /// <summary>
    /// The per-user uninstall entry Inno Setup writes (Marquee.iss's AppId
    /// plus "_is1"), compared with the folder this .exe runs from.
    /// </summary>
    private static bool DetectInstalledCopy()
    {
        try
        {
            using var key = Registry.CurrentUser.OpenSubKey(
                @"Software\Microsoft\Windows\CurrentVersion\Uninstall\{879059C2-E41A-4B11-9541-A33DABB34E0A}_is1");
            if (key?.GetValue("InstallLocation") is not string location || location.Length == 0)
            {
                return false;
            }
            static string Normalize(string path) => Path.GetFullPath(path).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            return string.Equals(Normalize(location), Normalize(AppContext.BaseDirectory), StringComparison.OrdinalIgnoreCase);
        }
        catch (Exception error) when (error is System.Security.SecurityException or UnauthorizedAccessException or IOException or ArgumentException)
        {
            return false;
        }
    }

    /// <summary>Asks GitHub for the newest release. Does nothing while a check or install is under way.</summary>
    public async Task CheckAsync()
    {
        if (IsBusy)
        {
            return;
        }
        Phase = UpdatePhase.Checking;
        ErrorMessage = null;
        lastCheckKind = null;
        try
        {
            var result = await service.CheckAsync(AppVersion.Current);
            LatestRelease = result.Latest;
            // An install started while this was out; its phase wins.
            if (Phase != UpdatePhase.Checking)
            {
                return;
            }
            Update = result.Update;
            Phase = result.IsUpToDate ? UpdatePhase.UpToDate : UpdatePhase.Available;
        }
        catch (UpdateException error)
        {
            lastCheckKind = error.Kind;
            if (Phase != UpdatePhase.Checking)
            {
                return;
            }
            // A known update stays on offer when a later check fails.
            if (Update != null)
            {
                Phase = UpdatePhase.Available;
            }
            else
            {
                ErrorMessage = error.Message;
                Phase = UpdatePhase.Failed;
            }
        }
    }

    /// <summary>
    /// Downloads the installer, checks it, starts it and quits. Checks first
    /// when no update is known yet (a click before the first check finished).
    /// </summary>
    public async Task InstallAsync()
    {
        if (IsBusy || !IsInstalledCopy)
        {
            return;
        }
        if (Update == null)
        {
            await CheckAsync();
        }
        if (Update is not { } available)
        {
            return;
        }
        installCancellation?.Dispose();
        installCancellation = new CancellationTokenSource();
        var token = installCancellation.Token;
        ErrorMessage = null;
        Progress = 0;
        Phase = UpdatePhase.Downloading;
        try
        {
            var expected = await service.ExpectedSha256Async(available, token);
            var installer = Path.Combine(DownloadFolder, $"Marquee-Setup-{available.Version}.exe");
            var reporter = new global::System.Progress<double>(value => Progress = value);
            await service.DownloadAsync(available, expected, installer, reporter, token);

            Phase = UpdatePhase.Installing;
            Launch(installer);
            // Give the progress line a moment to say what's happening, then
            // get out of the installer's way.
            await Task.Delay(TimeSpan.FromMilliseconds(600), CancellationToken.None);
            Microsoft.UI.Xaml.Application.Current.Exit();
        }
        catch (UpdateException error)
        {
            ErrorMessage = error.Message;
            Phase = UpdatePhase.Failed;
        }
        catch (OperationCanceledException)
        {
            Phase = UpdatePhase.Available;
        }
    }

    /// <summary>Stops a download in progress.</summary>
    public void CancelInstall() => installCancellation?.Cancel();

    private static void Launch(string installer)
    {
        try
        {
            var started = Process.Start(new ProcessStartInfo(installer)
            {
                Arguments = InstallerArguments,
                UseShellExecute = false,
                WorkingDirectory = Path.GetDirectoryName(installer) ?? Path.GetTempPath(),
            });
            if (started == null)
            {
                throw new UpdateException(UpdateErrorKind.LaunchFailed);
            }
        }
        catch (Exception error) when (error is Win32Exception or InvalidOperationException or IOException)
        {
            throw new UpdateException(UpdateErrorKind.LaunchFailed, error);
        }
    }
}

File: mac/Marquee/App/Updates/Updater.swift
import AppKit
import Observation
import OSLog

/// Keeps this Mac on the newest Marquee: checks GitHub's latest release ten
/// seconds after launch and then daily (and from Marquee › Check for
/// Updates…), and on "Update" downloads it, checks it, and swaps it in, then
/// quits so the new version opens in its place (`UpdateService`,
/// `UpdateInstaller`).
@MainActor
@Observable
final class Updater {
    enum Phase: Equatable {
        case idle
        case checking
        case upToDate
        /// `update` is newer than this app.
        case available
        /// 0...1 of the zip.
        case downloading(Double)
        /// Unpacking and checking the app.
        case verifying
        /// The swap is running; Marquee is about to quit.
        case relaunching
        case failed(Failure)
    }

    struct Failure: Equatable {
        let error: UpdateError
        /// The checked new app, kept in Downloads when it couldn't be put in
        /// place (for "Show in Finder").
        var downloadedApp: URL?
    }

    enum CheckResult: Equatable {
        case upToDate(latest: AppVersion?)
        case available(AvailableUpdate)
        case failed(UpdateError)
    }

    static let firstCheckDelay: Duration = .seconds(10)
    static let checkInterval: Duration = .seconds(24 * 60 * 60)
    /// The version whose "is available" banner has been shown.
    static let announcedKey = "marquee.update.announcedVersion"

    private(set) var phase: Phase = .idle
    /// The newest release, when it's newer than this app.
    private(set) var update: AvailableUpdate?
    /// Why the last check didn't get an answer (shown in Settings › About).
    private(set) var checkError: UpdateError?

    let currentVersion: AppVersion?

    /// A newer release was found by a check: once per version, the main
    /// window's banner says so.
    @ObservationIgnored var onNewUpdate: ((AvailableUpdate) -> Void)?

    @ObservationIgnored private let service: UpdateService
    @ObservationIgnored private let defaults: UserDefaults
    @ObservationIgnored private var scheduleTask: Task<Void, Never>?
    @ObservationIgnored private var installTask: Task<Void, Never>?

    private static let logger = Logger(subsystem: "com.timmyamant.Marquee", category: "updates")

    init(
        service: UpdateService = UpdateService(),
        currentVersion: AppVersion? = .current,
        defaults: UserDefaults = .standard
    ) {
        self.service = service
        self.currentVersion = currentVersion
        self.defaults = defaults
    }

    var isInstalling: Bool {
        switch phase {
        case .downloading, .verifying, .relaunching: return true
        default: return false
        }
    }

    // MARK: Checking

    func startAutomaticChecks() {
        guard scheduleTask == nil else { return }
        scheduleTask = Task { [weak self] in
            try? await Task.sleep(for: Updater.firstCheckDelay)
            while !Task.isCancelled {
                await self?.check()
                try? await Task.sleep(for: Updater.checkInterval)
            }
        }
    }

    /// Asks GitHub for the latest release. A failure leaves an update that
    /// was already found in place.
    @discardableResult
    func check() async -> CheckResult {
        guard !isInstalling else { return update.map(CheckResult.available) ?? .upToDate(latest: currentVersion) }
        let previous = phase
        if update == nil { phase = .checking }

        let result: CheckResult
        do {
            let release = try await service.latestRelease()
            guard let latest = AppVersion(release.tagName) else { throw UpdateError.unreadableRelease }
            if let currentVersion, latest <= currentVersion {
                result = .upToDate(latest: latest)
            } else if let available = AvailableUpdate(release: release) {
                result = .available(available)
            } else {
                // Newer, but CI hasn't attached the Mac download yet.
                throw UpdateError.noDownload
            }
        } catch {
            result = .failed(error as? UpdateError ?? .unreachable)
        }

        guard !isInstalling else { return result }
        switch result {
        case let .upToDate(latest):
            Self.logger.info("Up to date (latest release \(latest?.description ?? "?", privacy: .public))")
            checkError = nil
            update = nil
            phase = .upToDate
        case let .available(available):
            Self.logger.info("Marquee \(available.version.description, privacy: .public) is available")
            checkError = nil
            update = available
            phase = .available
            announceOnce(available)
        case let .failed(error):
            Self.logger.info("Update check failed: \(error.localizedDescription, privacy: .public)")
            checkError = error
            phase = update == nil ? (previous == .checking ? .idle : previous) : .available
        }
        return result
    }

    private func announceOnce(_ available: AvailableUpdate) {
        guard defaults.string(forKey: Self.announcedKey) != available.version.description else { return }
        defaults.set(available.version.description, forKey: Self.announcedKey)
        onNewUpdate?(available)
    }

    // MARK: Installing

    /// "Update": download, check, swap, relaunch. Progress and any failure
    /// show in `phase`.
    func install() {
        guard let update, !isInstalling else { return }
        phase = .downloading(0)
        installTask = Task { [weak self] in
            await self?.install(update)
        }
    }

    private func install(_ update: AvailableUpdate) async {
        let bundle = Bundle.main.bundleURL
        let blocker = UpdateInstaller.blocker(for: bundle)
        do {
            let workspace = try UpdateInstaller.workspace(near: blocker == nil ? bundle : nil)
            let sha256 = try await service.expectedSHA256(for: update)
            let zip = try await service.download(update, sha256: sha256, into: workspace) { [weak self] fraction in
                Task { @MainActor in self?.downloaded(fraction) }
            }
            phase = .verifying
            let app = try await UpdateInstaller.unpack(zip, in: workspace, expecting: update.version)

            if let blocker {
                // Checked and ready, but it can't go where this copy is.
                let kept = UpdateInstaller.keepForManualInstall(app, version: update.version)
                phase = .failed(Failure(error: blocker, downloadedApp: kept))
                return
            }

            phase = .relaunching
            // Long enough to read "Marquee will quit and reopen".
            try? await Task.sleep(for: .seconds(1.5))
            try UpdateInstaller.launchSwap(current: bundle, new: app, pid: ProcessInfo.processInfo.processIdentifier, workspace: workspace)
            Self.logger.notice("Installing Marquee \(update.version.description, privacy: .public); quitting so it can open")
            NSApp.terminate(nil)
        } catch {
            let failure = error as? UpdateError ?? .downloadFailed
            Self.logger.error("Update failed: \(failure.localizedDescription, privacy: .public)")
            phase = .failed(Failure(error: failure))
        }
    }

    private func downloaded(_ fraction: Double) {
        guard case let .downloading(shown) = phase, fraction - shown >= 0.01 || fraction >= 1 else { return }
        phase = .downloading(fraction)
    }
}
